Objects/fileobject.c

```diff
@@ -119,7 +119,8 @@ file_PyObject_Print(PyObject *op, PyFileObject *f, int flags)
 static PyFileObject*
 dircheck(PyFileObject* f)
 {
-#if defined(HAVE_FSTAT) && defined(S_IFDIR) && defined(EISDIR)
+#if !defined(MS_WINDOWS) && defined(HAVE_FSTAT) && defined(S_IFDIR) && \
```

Comment:
Yes, it's possible if we don't want to make `dircheck` work on Windows (I changed `dircheck` instead of its callers just because it makes a smaller patch). It also makes sense to put `dircheck` under `#ifndef MS_WINDOWS` in this case.

If `dircheck` is to be fixed instead, the call in `fill_file_fields` can't be removed because `fill_file_fields` is called by `PyFile_FromFile` which accepts `FILE *` of potentially unknown origin (not necessarily from `fopen`).

I may update the patch once there is a decision about fixing `dircheck` for Windows.

Reply:
For `PyFile_FromFile` I'm not really concerned about a C extension doing something silly. I don't think the check is necessary. It never worked, and no one ever complained that it didn't work, so my vote is to retain the status quo by either skipping the `dircheck` call or making it a no-op.

Reply:
Now I see your points here and [above](https://github.com/python/cpython/pull/9520#discussion_r220675393), but is there any reason to make `os.fdopen` behave differently compared to `PyFile_FromFile` (and, as a result, `stdin` check)? Directory checks never worked in both of them.

Reply:
My reasoning was that we can end up with an fd for a directory using PyWin32 and `msvcrt.open_osfhandle`. OTOH, C `FILE` streams are definitely outside of the domain of Python. If we consider code that directly calls `CreateFile` (via PyWin32 or ctypes) to be on the periphery, then the check in `os.fdopen` doesn't need to be fixed. It never worked, so the status quo is preserved by skipping it.

Reply:
Thank you, I understand your reasoning. A nitpick: we can still abuse  `PyFile_FromFile` without a third-party extension module since it's used to wrap std streams.
```python
import ctypes, msvcrt, subprocess, sys
h = ctypes.windll.kernel32.CreateFileA(
  ".",
  0,
  0,
  0,
  3,          # OPEN_EXISTING
  0x02000000, # FILE_FLAG_BACKUP_SEMANTICS 
  0,
)
fd = msvcrt.open_osfhandle(h, 0)
subprocess.check_call([sys.executable, '-c', 'import sys; sys.stdout.write("OK")'],
                      stdout=fd)        
```
The above prints the following:
```
close failed in file object destructor:
sys.excepthook is missing
lost sys.stderr
```

Personally, I'd either fix both `os.fdopen` and `PyFile_FromFile` or preserve the status quo, and I'm more inclined to the latter.